File: Data Structures & Algorithms/max-area-of-island/submission-1.js
class Solution {
  /**
   * @param {number[][]} grid
   * @return {number}
   */
  maxAreaOfIsland(grid) {
    if (grid.length === 0) return 0;
    const rows = grid.length;
    const cols = grid[0].length;
    const visited = Array.from({ length: rows }, () =>
      new Array(cols).fill(false),
    );
    let maxArea = 0;
    for (let i = 0; i < rows; i++) {
      for (let j = 0; j < cols; j++) {
        if (grid[i][j] === 1 && visited[i][j] === false) {
          const area = this.dfs(i, j, rows, cols, visited, grid);
          maxArea = Math.max(maxArea, area);
        }
      }
    }
    return maxArea;
  }

  dfs(i, j, rows, cols, visited, grid) {
    if (
      i < 0 ||
      j < 0 ||
      i >= rows ||
      j >= cols ||
      visited[i][j] === true ||
      grid[i][j] === 0
    )
      return 0;
    visited[i][j] = true;
    let area = 1;
    area += this.dfs(i - 1, j, rows, cols, visited, grid); // up
    area += this.dfs(i + 1, j, rows, cols, visited, grid); // down
    area += this.dfs(i, j - 1, rows, cols, visited, grid); // left
    area += this.dfs(i, j + 1, rows, cols, visited, grid); // right
    return area;
  }
}

File: Data Structures & Algorithms/count-paths/submission-1.java
class Solution {
     public int uniquePaths(int m, int n) {
        int[][] memo =  new int[m][n];
        for (int[] mm : memo){
            Arrays.fill(mm,-1);
        }
        return uniquePathsRec(0,0,m,n,memo);
    }

    private int uniquePathsRec(int i,int j,int m,int n,int[][] memo){
        if(i >= m || j >= n) return 0;
        if(i == m-1 && j == n-1) return 1;
        if(memo[i][j] != -1) return memo[i][j];
        var left = uniquePathsRec(i+1,j,m,n,memo);
        var right = uniquePathsRec(i,j+1,m,n,memo);
        memo[i][j] = left+ right;
        return memo[i][j];
    }
}
